File: src/format/javadoc/char-stream.ts
// Port of google-java-format core/.../java/javadoc/CharStream.java.
//
// String reader for the lexer: tryConsume* match at the current position, then
// readAndResetRecorded() returns and consumes the match. Regexes must be anchored
// (we use the sticky flag `y` to match only at the current position).

export class CharStream {
  private position = 0;
  private tokenEnd = -1; // negative => no pending token

  constructor(private readonly input: string) {}

  tryConsume(expected: string): boolean {
    if (!this.input.startsWith(expected, this.position)) return false;
    this.tokenEnd = this.position + expected.length;
    return true;
  }

  /** `pattern` must be a sticky (`y`) RegExp so it matches only at `position`. */
  tryConsumeRegex(pattern: RegExp): boolean {
    pattern.lastIndex = this.position;
    const m = pattern.exec(this.input);
    if (!m || m.index !== this.position) return false;
    this.tokenEnd = this.position + m[0].length;
    return true;
  }

  readAndResetRecorded(): string {
    const result = this.input.slice(this.position, this.tokenEnd);
    this.position = this.tokenEnd;
    this.tokenEnd = -1;
    return result;
  }

  isExhausted(): boolean {
    return this.position === this.input.length;
  }

  pos(): number {
    return this.position;
  }
}

File: togo/internal/format/javadoc/char_stream.go
// Port of src/format/javadoc/char-stream.ts.
//
// String reader for the lexer. Regexes must be anchored at the cursor; we build
// them with a leading `\A`-equivalent via FindStringIndex from the offset.

package javadoc

import "regexp"

type charStream struct {
	input    string
	position int
	tokenEnd int
}

func newCharStream(input string) *charStream {
	return &charStream{input: input, tokenEnd: -1}
}

func (c *charStream) tryConsume(expected string) bool {
	if c.position+len(expected) > len(c.input) || c.input[c.position:c.position+len(expected)] != expected {
		return false
	}
	c.tokenEnd = c.position + len(expected)
	return true
}

// tryConsumeRegex matches `pattern` only at the current position. `pattern` must
// be anchored with `^` and compiled normally; we slice from the cursor so `^`
// binds there.
func (c *charStream) tryConsumeRegex(pattern *regexp.Regexp) bool {
	loc := pattern.FindStringIndex(c.input[c.position:])
	if loc == nil || loc[0] != 0 {
		return false
	}
	c.tokenEnd = c.position + loc[1]
	return true
}

func (c *charStream) readAndResetRecorded() string {
	result := c.input[c.position:c.tokenEnd]
	c.position = c.tokenEnd
	c.tokenEnd = -1
	return result
}

func (c *charStream) isExhausted() bool { return c.position == len(c.input) }
